File: 2825-make-string-a-subsequence-using-cyclic-increments/2825-make-string-a-subsequence-using-cyclic-increments.py
class Solution:
    def canMakeSubsequence(self, str1: str, str2: str) -> bool:
        r = 0
        for l, lit in enumerate(str1):
            if r == len(str2):
                break
            rit = str2[r]
            if self._sim(lit, rit):
                r += 1
        return r == len(str2)
    
    def _dec(self, s: str):
        chars = [chr(ord(it) - 1) if it != 'a' else 'z' for it in s]
        return ''.join(chars)

    def _sim(self, lhs, rhs):
        return lhs in [rhs, self._dec(rhs)]
        

File: 2825-make-string-a-subsequence-using-cyclic-increments/2825-make-string-a-subsequence-using-cyclic-increments.c
char increased(char ch);
bool similar(char lhs, char rhs);

bool canMakeSubsequence(char* str1, char* str2) {
    char* rit = str2;

    for (char* lit = str1; *lit && *rit; ++lit) {
        if (similar(*lit, *rit)) {
            ++rit;
        }
    }

    return *rit == '\0';
}

char increased(char ch) {
    int idx = ch - 'a';
    return (idx + 1) % 26 + 'a';
}

bool similar(char lhs, char rhs) {
    return lhs == rhs || increased(lhs) == rhs;
}